shared-bindings/adcbuffer/BufferedInput.c

```diff
@@ -0,0 +1,180 @@
+    uint8_t bytes_per_sample = 1;
+
+    // Bytes Per Sample
+    if (bufinfo.typecode == 'h' || bufinfo.typecode == 'H') {
+        bytes_per_sample = 2;
+    } else if (bufinfo.typecode != 'b' && bufinfo.typecode != 'B' && bufinfo.typecode != BYTEARRAY_TYPECODE) {
+        mp_raise_ValueError(translate("sample_source buffer must be a bytearray or array of type 'h', 'H', 'b' or 'B'"));
+    }
+
+    // Validate sample rate here
+    uint32_t sample_rate = (uint32_t)mp_arg_validate_int_range(args[ARG_sample_rate].u_int, 1, 500000, MP_QSTR_sample_rate);
+
+    // Create local object
+    adcbuffer_bufferedinput_obj_t *self = m_new_obj(adcbuffer_bufferedinput_obj_t);
+    self->base.type = &adcbuffer_bufferedinput_type;
+
+    // Call local intereface in ports/common-hal/adcbuffer
+    common_hal_adcbuffer_bufferedinput_construct(self,
+        pin,
+        ((uint8_t *)bufinfo.buf),
+        bufinfo.len,
+        bytes_per_sample,
+        signed_samples,
+        sample_rate
+        );
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|     def deinit(self) -> None:
+//|         """Turn off the BufferedInput and release the pin for other use."""
+//|         ...
+//|
+STATIC mp_obj_t adcbuffer_bufferedinput_deinit(mp_obj_t self_in) {
+    adcbuffer_bufferedinput_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_adcbuffer_bufferedinput_deinit(self);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(adcbuffer_bufferedinput_deinit_obj, adcbuffer_bufferedinput_deinit);
+
+STATIC void check_for_deinit(adcbuffer_bufferedinput_obj_t *self) {
+    if (common_hal_adcbuffer_bufferedinput_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+//|  Provided by context manager helper.
+//|
+//|     def __exit__(self) -> None:
+//|         """Automatically deinitializes the hardware when exiting a context. See
+//|         :ref:`lifetime-and-contextmanagers` for more info."""
+//|         ...
+//|
+STATIC mp_obj_t adcbuffer_bufferedinput___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_adcbuffer_bufferedinput_deinit(args[0]);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(adcbuffer_bufferedinput___exit___obj, 4, 4, adcbuffer_bufferedinput___exit__);
+
+//|     value: None
+//|         """Fills the supplied buffer with ADC values using DMA transfer.
+//|     If the buffer is 8-bit, then values are 8-bit shifted and error bit is off.
+//|     If buffer is 16-bit, then values are not shifted and error bit is present.
+//|     Number of transfers is always the number of samples which is the array
+//|     byte length divided by the bytes_per_sample."""
+//|
+STATIC mp_obj_t adcbuffer_bufferedinput_obj_readmultiple(mp_obj_t self_in) {
+    adcbuffer_bufferedinput_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    check_for_deinit(self);
+    common_hal_adcbuffer_bufferedinput_readmultiple(self);
+    return mp_const_none;
```

Comment:
`readmultiple()` takes no arguments, but it could take an optional count of the number of samples to read. If the count is omitted, it would fill the buffer. But it might be convenient to be able to read a smaller number of samples into a larger reused buffer.

Reply:
Did change the interface from readmultiple() to read(). I agree. if no arguments, then fill the buffer using buffer length. Otherwise use the argument checking to make sure the unsigned integer is no greater than buffer length. This should not affect speed or operations in most cases. Buffer initialization WILL BE left to user for max speed and flexibility.